bootloader/piboot.go

```diff
@@ -0,0 +1,395 @@
+	fullCmdline := env.Get("snapd_full_cmdline_args")
+	if fullCmdline != "" {
+		base = fullCmdline
+	} else {
+		buf, err := ioutil.ReadFile(defaultsFile)
+		if err != nil {
+			return err
+		}
+
+		lines := strings.Split(string(buf), "\n")
+		base = lines[0]
+		base += " " + env.Get("snapd_extra_cmdline_args")
+	}
+
+	mode := env.Get("snapd_recovery_mode")
+	cmdline := base + " snapd_recovery_mode=" + mode
+	if mode != "run" {
+		cmdline += " snapd_recovery_system=" + env.Get("snapd_recovery_system")
+	}
+	// Signal when we are trying a new kernel
+	kernelStatus := env.Get("kernel_status")
+	if kernelStatus == "try" {
+		cmdline += " kernel_status=trying"
+	}
+	cmdline += "\n"
+
+	logger.Debugf("writing kernel command line to %s", outFile)
+
+	return osutil.AtomicWriteFile(outFile, []byte(cmdline), 0644, 0)
+}
+
+// Configure pi bootloader with a given os_prefix. cfgDir contains the
+// config files, and dstDir is where we will place the kernel command
+// line.
+func (p *piboot) applyConfig(env *ubootenv.Env,
+	configFile, prefix, cfgDir, dstDir string) error {
+
+	kernStat := env.Get("kernel_status")
+	cmdlineFile := filepath.Join(dstDir, "cmdline.txt")
+	// These two files are part of the gadget
+	templCmdlineFile := filepath.Join(cfgDir, "cmdline.templ.txt")
+	templConfigFile := filepath.Join(cfgDir, "config.templ.txt")
+
+	if err := p.createCmdline(env, templCmdlineFile, cmdlineFile); err != nil {
+		return err
+	}
+	if err := p.setRPiCfgOsPrefix(prefix, templConfigFile,
+		filepath.Join(cfgDir, configFile)); err != nil {
+		return err
+	}
+
+	if kernStat == "try" {
+		// The reboot parameter makes sure we use tryboot.cfg config
+		// TODO Maybe this should be done somewhere else in snapd?
+		if err := osutil.AtomicWriteFile("/run/systemd/reboot-param",
+			[]byte("0 tryboot\n"), 0644, 0); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *piboot) GetBootVars(names ...string) (map[string]string, error) {
+	out := make(map[string]string)
+
+	env, err := ubootenv.OpenWithFlags(p.envFile(), ubootenv.OpenBestEffort)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, name := range names {
+		out[name] = env.Get(name)
+	}
+
+	return out, nil
+}
+
+func (p *piboot) InstallBootConfig(gadgetDir string, blOpts *Options) error {
+	gadgetFile := filepath.Join(gadgetDir, pibootCfgFilename)
+	st, err := os.Stat(gadgetFile)
+	if err != nil {
+		return err
+	}
+	if st.Size() == 0 {
+		// We create a valid env file if the one from gadget is empty
+		err := os.MkdirAll(filepath.Dir(p.envFile()), 0755)
+		if err != nil {
+			return err
+		}
+
+		// TODO: what's a reasonable size for this file?
+		env, err := ubootenv.Create(p.envFile(), 4096)
+		if err != nil {
+			return err
+		}
+
+		if err := env.Save(); err != nil {
+			return nil
+		}
+
+		return nil
+	}
+
+	systemFile := p.envFile()
+	return genericInstallBootConfig(gadgetFile, systemFile)
+}
+
+func (p *piboot) layoutKernelAssetsToDir(snapf snap.Container, dstDir string) error {
```

Comment:
how atomic does this need to be or can be?

Reply:
I don't think it needs to be atomic, as the critical point is when `tryboot.txt` is written with an `os_prefix` that points to the extracted kernel assets.

Reply:
On top of what Alfonso said, the critical bit for after having successfully booted a try kernel will be writing a config.txt file with the right value of os_prefix to point to the new one we extracted, since we don't have to move around any existing files we just write the new value like we would do with this function, so there are no critical sections on a MarkBootSuccessful run as well so I agree that this does not need to be atomic